### PR TITLE
[bug][train] Fix max_seq_len calculation

### LINKED ISSUE #1154: [bug][train] `max_seq_len` auto-calculation is incorrect for multi-turn in `seq_mean_token_sum_norm`
### Problem

When `trainer.algorithm.max_seq_len` is not explicitly set, `validate_cfg()` computes it as:

```python
max_seq_len = cfg.generator.max_input_length + cfg.generator.sampling_params.max_generate_length
```

This is incorrect for multi-turn conversations because `max_generate_length` is a **per-turn** `max_tokens` parameter passed to the inference engine for each generation call, not the total response token budget across all turns.

For example, with `max_turns=5`, `max_input_length=2048`, and `max_generate_length=4096`:
- **Current auto-calculation**: `max_seq_len = 2048 + 4096 = 6144`
- **Actual possible sequence length**: up to `2048 + 5 * 4096 = 22528` (prompt + 5 turns of generation, plus tool outputs/observations between turns)

This means the `seq_mean_token_sum_norm` (Dr. GRPO) loss reduction normalizes by a value that is too small for multi-turn, which would scale up the loss magnitude for multi-turn sequences relative to what Dr. GRPO intends.

### Where it happens

`skyrl/train/utils/utils.py` in `validate_cfg()` (and the mirrored `skyrl-train/skyrl_train/utils/utils.py`):

```python
if cfg.trainer.algorithm.max_seq_len is None:
    # TODO(Charlie): This calculation is not correct for multi-turn...
    if isinstance(cfg, DictConfig):
        new_cfg = OmegaConf.create(cfg.trainer.algorithm)
        new_cfg.max_seq_len = cfg.generator.max_input_length + cfg.generator.sampling_params.max_generate_length
        cfg.trainer.algorithm = new_cfg
    else:
        cfg.trainer.algorithm.max_seq_len = (
            cfg.generator.max_input_length + cfg.generator.sampling_params.max_generate_length
        )
```

### Current workaround

Users can set `trainer.algorithm.max_seq_len` explicitly (https://github.com/NovaSky-AI/SkyRL/pull/1153). For Harbor, this is already done via `trainer.algorithm.max_seq_len=$MAX_MODEL_LEN` in the example scripts.

### Possible fixes

1. **Require explicit `max_seq_len` when `loss_reduction=seq_mean_token_sum_norm`** — remove the auto-calculation entirely and error if not set. This is the safest option since the "correct" value depends on the use case.
2. **Use the vLLM engine's `max_model_len`** as the default, since that represents the true context window limit.
3. **Factor in `max_turns`** in the auto-calculation, e.g. `max_input_length + max_turns * max_generate_length`. This is still an approximation since tool outputs between turns add tokens too, but it's closer to correct.

Option 1 seems cleanest — the auto-calculation was always a convenience heuristic, and for the loss reduction that actually uses this value, users should think about what the right normalization constant is for their setup.

### PR DESCRIPTION
**Pull request by tamoghnokandar**
Fixes #1154 
## Summary

This PR removes the implicit `max_seq_len` heuristic calculation and requires users to set it explicitly when using `trainer.algorithm.loss_reduction=seq_mean_token_sum_norm`.

## Changes

- Removed the automatic `trainer.algorithm.max_seq_len` default from `SkyRLTrainConfig.__post_init__`
- Added config validation that requires `trainer.algorithm.max_seq_len` to be explicitly set when `loss_reduction == "seq_mean_token_sum_norm"`
- Updated config comments and docs to reflect the new behavior and explain that `max_seq_len` must be chosen based on the user’s intended sequence-length normalization budget
- Updated tests to cover:
  - `max_seq_len` remaining `None` by default
  - explicit `max_seq_len` values being preserved
  - `validate_cfg()` failing when `seq_mean_token_sum_norm` is used without `max_seq_len`
  - `validate_cfg()` continuing to allow `token_mean` and `sequence_mean` without `max_seq_len`
  - `validate_cfg()` passing when `seq_mean_token_sum_norm` is used with an explicit `max_seq_len`

## Testing

- Updated `tests/train/test_config.py` for the new behavior

---

<a href="https://app.devin.ai/review/novasky-ai/skyrl/pull/1303" target="_blank">
  <picture>
    <source media="(prefers-color-scheme: dark)" srcset="https://static.devin.ai/assets/gh-open-in-devin-review-dark.svg?v=1">
    Open with Devin
  </picture>
</a>

### CODE REVIEW

**Review by SumanthRH — CHANGES_REQUESTED** [review]
@tamoghnokandar can you merge the latest changes from main? We've had some important updates, especially https://github.com/NovaSky-AI/SkyRL/commit/bf243b8d64bf5afd0c64ccd8b6af6d2285b800be

**Comment by tamoghnokandar** [comment]
Done!

File: skyrl/train/utils/utils.py [hunk 1]
```diff
@@ -276,6 +276,14 @@ def validate_cfg(cfg: SkyRLTrainConfig):
         f"invalid loss_reduction: {cfg.trainer.algorithm.loss_reduction}. "
         f"Must be one of `['token_mean', 'sequence_mean', 'seq_mean_token_sum_norm']`"
     )
+    if cfg.trainer.algorithm.loss_reduction == "seq_mean_token_sum_norm":
+        if cfg.trainer.algorithm.max_seq_len is None:
+            raise ValueError(
+                "`trainer.algorithm.max_seq_len` must be set explicitly when "
+                "`trainer.algorithm.loss_reduction='seq_mean_token_sum_norm'`. "
+                "Choose the total sequence-length normalization constant for your setup; "
+                "this often matches the model context window / vLLM `max_model_len` when appropriate."
```
[the remark this answers is not in the document]
Comment (SumanthRH) [reply skyrl/train/utils/utils.py:292]: @tamoghnokandar this is important - can you grep for all usages of `seq_mean_token_sum_norm` in our example scripts and ensure that `max_seq_len` is explicitly passed in now (calculate it based on the generation and input lengths in the script).

[the remark this answers is not in the document]
Comment (SumanthRH) [reply skyrl/train/utils/utils.py:292]: https://github.com/NovaSky-AI/SkyRL/blob/1f8c6f24bc8ae70e08e9b1bee85006980ba8fff3/examples/train/algorithms/drgrpo/run_drgrpo_gsm8k.sh#L15

https://github.com/NovaSky-AI/SkyRL/blob/1f8c6f24bc8ae70e08e9b1bee85006980ba8fff3/examples/train_integrations/harbor/run_codecontest.sh#L38

https://github.com/NovaSky-AI/SkyRL/blob/1f8c6f24bc8ae70e08e9b1bee85006980ba8fff3/skyrl-agent/examples/run_skyrl/run_skyrl_swe.sh#L67

File: skyrl/train/config/config.py
Comment (SumanthRH) [inline skyrl/train/config/config.py:1]: can you also add a note in the loss_type docstring that `max_seq_len` is required for `seq_man_token_sum_norm` now?

https://github.com/tamoghnokandar/SkyRL/blob/9c44257041101fe52893ad554bff026d49527b58/skyrl/train/config/config.py#L351

File: skyrl/train/config/config.py [hunk 2]
```diff
@@ -716,7 +716,6 @@ def __post_init__(self):
         # so workers can access it without needing the generator config
         if self.trainer.algorithm.temperature is None:
             self.trainer.algorithm.temperature = self.generator.sampling_params.temperature
-
         if self.trainer.algorithm.max_seq_len is None:
             # NOTE (erictang000): this is the max sequence length including the prompt, since max response length
             # per batch can be variable based on the prompt length. This is used to normalize the loss for
```
[the remark this answers is not in the document]
Comment (tamoghnokandar) [reply skyrl/train/config/config.py:721]: @SumanthRH Should I remove this? For seq_mean_token_sum_norm loss we are setting it explicitly anyways. This is not required for other loss functions right?

[the remark this answers is not in the document]
Comment (SumanthRH) [reply skyrl/train/config/config.py:721]: @tamoghnokandar yes remove it

**Comment by tamoghnokandar** [comment]
@SumanthRH Ready to be merged

**Review by SumanthRH — CHANGES_REQUESTED** [review]
One small fix: Can we ensure that we update this script as well now with an explicit `max_seq_len` argument? 

https://github.com/NovaSky-AI/SkyRL/blob/1f8c6f24bc8ae70e08e9b1bee85006980ba8fff3/skyrl-agent/examples/run_skyrl/run_skyrl_swe.sh#L67

**Review by CharlieFRuan — APPROVED** [review]
This is great, thank you!

File: examples/train_integrations/harbor/run_harbor_gen.sh [hunk 3]
```diff
@@ -37,6 +38,8 @@ uv run --isolated --extra fsdp --extra harbor -m examples.train_integrations.har
   generator.inference_engine.http_endpoint_host="127.0.0.1" \
   generator.inference_engine.http_endpoint_port=8000 \
   generator.sampling_params.max_generate_length=4096 \
+  trainer.algorithm.max_seq_len=$MAX_MODEL_LEN \
+  generator.inference_engine.engine_init_kwargs.max_model_len=$MAX_MODEL_LEN \
```
Comment (SumanthRH) [inline examples/train_integrations/harbor/run_harbor_gen.sh:42]: hmm why constrain inference engine max model len?

Comment (CharlieFRuan) [reply examples/train_integrations/harbor/run_harbor_gen.sh:42]: This is just a generation entrypoint, so it doesn't really matter. For training this is needed to limit the total context window size for the model, which should match `$MAX_MODEL_LEN`. For token-in-token-out, that is essentially the maximum sequence length (padded or not)

**Review by SumanthRH — APPROVED** [review]
